crypto/hpke/hpke.c

```diff
@@ -0,0 +1,2384 @@
+    } else {
+        erv = hpke_aead_dec(ctx->libctx, ctx->propq, ctx->suite,
+                            key, keylen, nonce, noncelen,
+                            aad, aadlen, ct, *ctlen, pt, ptlen);
+    }
+    if (erv != 1) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+
+err:
+    EVP_KDF_CTX_free(kctx);
+    return erv;
+}
+
+/* externally visible functions from below here */
+
+/**
+ * @brief contex creator
+ * @param mode is the desired HPKE mode
+ * @param suite specifies the KEM, KDF and AEAD to use
+ * @param libctx is the context to use
+ * @param propq is a properties string
+ * @return pointer to new context or NULL if error
+ */
+OSSL_HPKE_CTX *OSSL_HPKE_CTX_new(int mode, OSSL_HPKE_SUITE suite,
+                                 OSSL_LIB_CTX *libctx, const char *propq)
+{
+    OSSL_HPKE_CTX *ctx = NULL;
+
+    if (hpke_suite_check(suite) != 1)
+        return NULL;
+    if (mode < 0 || mode > OSSL_HPKE_MODE_PSKAUTH)
+        return NULL;
+    ctx = OPENSSL_zalloc(sizeof(OSSL_HPKE_CTX));
+    if (ctx == NULL)
+        return ctx;
+    ctx->libctx = libctx;
+    if (propq != NULL) {
+        ctx->propq = OPENSSL_strdup(propq);
+        if (ctx->propq == NULL)
+            goto err;
+    }
+    ctx->mode = mode;
+    ctx->suite = suite;
+    return ctx;
+
+err:
+    OSSL_HPKE_CTX_free(ctx);
+    return NULL;
+}
+
+/**
+ * @brief free up storage for a HPKE context
+ * @param ctx is the pointer to be free'd (can be NULL)
+ */
+void OSSL_HPKE_CTX_free(OSSL_HPKE_CTX *ctx)
+{
+    if (ctx == NULL)
+        return;
+    OPENSSL_free(ctx->propq);
+    if (ctx->exportersec)
+        OPENSSL_cleanse(ctx->exportersec, ctx->exporterseclen);
+    OPENSSL_free(ctx->exportersec);
+    OPENSSL_free(ctx->pskid);
+    OPENSSL_cleanse(ctx->psk, ctx->psklen);
+    OPENSSL_free(ctx->psk);
+    OPENSSL_free(ctx->shared_secret);
+    OPENSSL_free(ctx->ikme);
+
+    EVP_PKEY_free(ctx->authpriv);
+    EVP_PKEY_free(ctx->senderpriv);
+
+    OPENSSL_free(ctx->authpub);
+
+    OPENSSL_free(ctx);
+    return;
+}
+
+/**
+ * @brief set a PSK for an HPKE context
+ * @param ctx is the pointer for the HPKE context
+ * @param pskid is a string identifying the PSK
+ * @param psk is the PSK buffer
+ * @param psklen is the size of the PSK
+ * @return 1 for success, 0 for error
+ */
+int OSSL_HPKE_CTX_set1_psk(OSSL_HPKE_CTX *ctx,
+                           const char *pskid,
+                           const unsigned char *psk, size_t psklen)
+{
+    if (ctx == NULL || pskid == NULL || psk == NULL || psklen == 0) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    if (ctx->mode != OSSL_HPKE_MODE_PSK
+        && ctx->mode != OSSL_HPKE_MODE_PSKAUTH) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    /* free previous value if any */
+    OPENSSL_free(ctx->pskid);
+    OPENSSL_cleanse(ctx->psk, ctx->psklen);
+    OPENSSL_free(ctx->psk);
```

Comment:
There is a OPENSSL_clear_free()..

Reply:
Happy to do a pass that fixes that kind of thing more generally (was on my list anyway). Is there a doc somewhere that describes the pros/cons of that? If not I'd likely make an editing pass that does a cleanse/clear_free on anything resembling a secret but is that the correct approach? (I assume so, in which case I'll do that next day or so.)

Reply:
Use cleanse if the object is on the stack.. The clear_free just calls cleanse and then free. 
Anything that looks like a private/secret value should do the clear_free or cleanse (Taking care of the case of failures also that may have a temp value).

Reply:
Checking: I think if I put an IKM value into an OSSL_PARAM then I only need to clear_free or cleanse once?

Reply:
only caller needs to cleanse/clear_free